src/modules/deliveryoptimization/src/lib/DeliveryOptimization.c

```diff
@@ -0,0 +1,426 @@
+        if (MMI_OK != status)
+        {
+            if (JSONNumber == jsonPropertyType)
+            {
+                emptyJsonPayload = "0";
+            }
+            else if (JSONString == jsonPropertyType)
+            {
+                emptyJsonPayload = "\"\"";
+            }
+
+            *payloadSizeBytes = strlen(emptyJsonPayload);
+            *payload = (MMI_JSON_STRING)malloc(*payloadSizeBytes);
+
+            if (NULL != payload)
+            {
+                strncpy(*payload, emptyJsonPayload, *payloadSizeBytes);
+                status = MMI_OK;
+            }
+            else 
+            {
+                OsConfigLogError(DeliveryOptimizationGetLog(), "MmiGet failed to allocate %d bytes", *payloadSizeBytes + 1);
+                *payloadSizeBytes = 0;
+                status = ENOMEM;
+            }
+        }
+    }
+
+    if (IsFullLoggingEnabled())
+    {
+        OsConfigLogInfo(DeliveryOptimizationGetLog(), "MmiGet(%p, %s, %s, %.*s, %d) returning %d", clientSession, componentName, objectName, *payloadSizeBytes, *payload, *payloadSizeBytes, status);
+    }
+
+    return status;
+}
+
+int DeliveryOptimizationMmiSet(MMI_HANDLE clientSession, const char* componentName, const char* objectName, const MMI_JSON_STRING payload, const int payloadSizeBytes)
+{
+    int status = MMI_OK;
+    char *buffer = NULL;
+    JSON_Value* rootValue = NULL;
+    JSON_Object* rootObject = NULL;
+    JSON_Value* newValue = NULL;
+    JSON_Object* newObject = NULL;
+    unsigned int count = 0;
+    const char* name = NULL;
+    JSON_Value *currentValue = NULL;
+    const char* cacheHost = NULL;
+    int cacheHostSource = 0;
+    int cacheHostFallback = 0;
+    int percentageDownloadThrottle = 0;
+
+    if ((NULL == componentName) || (NULL == objectName) || (NULL == payload) || (payloadSizeBytes <= 0))
+    {
+        OsConfigLogError(DeliveryOptimizationGetLog(), "MmiSet(%s, %s, %p, %d) called with invalid arguments", componentName, objectName, payload, payloadSizeBytes);
+        status = EINVAL;
+    }
+    else if (!IsValidSession(clientSession))
+    {
+        OsConfigLogError(DeliveryOptimizationGetLog(), "MmiSet(%s, %s) called outside of a valid session", componentName, objectName);
+        status = EINVAL;
+    }
+    else if ((MMI_OK == status) && (strcmp(componentName, g_deliveryOptimizationComponentName)))
+    {
+        OsConfigLogError(DeliveryOptimizationGetLog(), "MmiSet called for an unsupported component name '%s'", componentName);
+        status = EINVAL;
+    }
+    else if ((MMI_OK == status) && (strcmp(objectName, g_desiredDeliveryOptimizationPoliciesObjectName)))
+    {
+        OsConfigLogError(DeliveryOptimizationGetLog(), "MmiSet called for an unsupported object name '%s'", objectName);
+        status = EINVAL;
+    }
+    else 
+    {
+        // Make sure that payload is null-terminated for json_parse_string.
+        buffer = malloc(payloadSizeBytes + 1);

```

Comment:
Perhaps use a local buffer than contains the JSON as a null terminated string and then for actual payload only copy payload size w/o null terminator. As it is payload will be null terminated and it must not be.

Reply:
Yes, adjusted to be more clear. During `MmiSet` it will copy the payload to a local buffer, with null terminator, so that it can be passed to `json_parse_string`. Thanks!